Src/AutoFixture.xUnit.net.UnitTest/FrozenAttributeTest.cs

```diff
@@ -82,18 +69,129 @@ public void GetCustomizationWithSpecificRegisteredTypeReturnsCorrectResult()
         {
             // Fixture setup
             var registeredType = typeof(string);
+#pragma warning disable 0618
             var sut = new FrozenAttribute { As = registeredType };
-            var parameter = typeof(TypeWithConcreteParameterMethod)
-                .GetMethod("DoSomething", new[] { typeof(ConcreteType) })
-                .GetParameters()
-                .Single();
+#pragma warning restore 0618
+            var parameter = AParameter<ConcreteType>();
             // Exercise system and verify outcome
             Assert.Throws<ArgumentException>(() => sut.GetCustomization(parameter));
             // Teardown
         }
+
+        [Fact]
+        public void GetCustomizationShouldMatchByExactParameterType()
+        {
+            // Fixture setup
+            var parameter = AParameter<object>();
+            var sut = new FrozenAttribute();
+            // Exercise system
+            var customization = sut.GetCustomization(parameter);
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            var exactTypeMatcher = matcher.Specifications.OfType<ExactTypeSpecification>().SingleOrDefault();
+            Assert.NotNull(exactTypeMatcher);
+            Assert.Equal(parameter.ParameterType, exactTypeMatcher.TargetType);
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationShouldMatchBySeedRequestForParameterType()
+        {
+            // Fixture setup
+            var parameter = AParameter<object>();
+            var sut = new FrozenAttribute();
+            // Exercise system
+            var customization = sut.GetCustomization(parameter);
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            var seedRequestMatcher = matcher.Specifications.OfType<SeedRequestSpecification>().SingleOrDefault();
+            Assert.NotNull(seedRequestMatcher);
+            Assert.Equal(parameter.ParameterType, seedRequestMatcher.TargetType);
+            // Teardown
+        }
+
+        [Fact]
+        public void GetCustomizationWithMatchingByDirectBaseTypeShouldMatchByBaseType()
+        {
+            // Fixture setup
+            var sut = new FrozenAttribute(Matching.DirectBaseType);
+            // Exercise system
+            var customization = sut.GetCustomization(AParameter<object>());
+            // Verify outcome
+            var freezer = Assert.IsAssignableFrom<FreezeOnMatchCustomization>(customization);
+            var matcher = Assert.IsType<OrRequestSpecification>(freezer.Matcher);
+            Assert.NotEmpty(matcher.Specifications.OfType<DirectBaseTypeSpecification>());

```

Comment:
There may be other tests below that drive towards the appropriate behaviour, so forgive me if the above comment is irrelevant, but I'm commenting as I go.